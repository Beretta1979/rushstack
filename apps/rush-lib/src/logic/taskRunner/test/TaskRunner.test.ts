import { EOL } from 'os';
import { TaskRunner } from '../TaskRunner';
import { ITaskWriter } from '@microsoft/stream-collator';
import { TaskStatus } from '../TaskStatus';
import { ITaskDefinition } from '../ITask';
import { StringBufferTerminalProvider, Terminal } from '@microsoft/node-core-library';

function createDummyTask(name: string, action?: () => void): ITaskDefinition {
  return {
    name,
    isIncrementalBuildAllowed: false,
    execute: (writer: ITaskWriter) => {
      if (action) {
        action();
      }
      return Promise.resolve(TaskStatus.Success);
    },
    hadEmptyScript: false
  };
}

function checkConsoleOutput(terminalProvider: StringBufferTerminalProvider): void {
  expect(terminalProvider.getOutput()).toMatchSnapshot();
  expect(terminalProvider.getVerbose()).toMatchSnapshot();
  expect(terminalProvider.getWarningOutput()).toMatchSnapshot();
  expect(terminalProvider.getErrorOutput()).toMatchSnapshot();
}

describe('TaskRunner', () => {
  let terminalProvider: StringBufferTerminalProvider;
  let terminal: Terminal;
  let taskRunner: TaskRunner;

  beforeEach(() => {
    terminalProvider = new StringBufferTerminalProvider(true);
    terminal = new Terminal(terminalProvider);
  });

  describe('Constructor', () => {
    it('throwsErrorOnInvalidParallelism', () => {
      expect(() => new TaskRunner({
        quietMode: false,
        parallelism: 'tequila',
        changedProjectsOnly: false,
        terminal,
        allowWarningsInSuccessfulBuild: false
      })).toThrowErrorMatchingSnapshot();
    });
  });

  describe('Dependencies', () => {
    beforeEach(() => {
      taskRunner = new TaskRunner({
          quietMode: false,
          parallelism: '1',
          changedProjectsOnly: false,
          terminal,
          allowWarningsInSuccessfulBuild: false
      });
    });

    it('throwsErrorOnNonExistentTask', () => {
      expect(() => taskRunner.addDependencies('foo', []))
        .toThrowErrorMatchingSnapshot();
    });

    it('throwsErrorOnNonExistentDependency', () => {
      taskRunner.addTask(createDummyTask('foo'));
      expect(() => taskRunner.addDependencies('foo', ['bar']))
        .toThrowErrorMatchingSnapshot();
    });

    it('detectsDependencyCycle', () => {
      taskRunner.addTask(createDummyTask('foo'));
      taskRunner.addTask(createDummyTask('bar'));
      taskRunner.addDependencies('foo', ['bar']);
      taskRunner.addDependencies('bar', ['foo']);
      expect(() => taskRunner.execute()).toThrowErrorMatchingSnapshot();
    });

    it('respectsDependencyOrder', () => {
      const result: Array<string> = [];
      taskRunner.addTask(createDummyTask('two', () => result.push('2')));
      taskRunner.addTask(createDummyTask('one', () => result.push('1')));
      taskRunner.addDependencies('two', ['one']);
      return taskRunner
        .execute()
        .then(() => {
          expect(result.join(',')).toEqual('1,2');
          checkConsoleOutput(terminalProvider);
        })
        .catch(error => fail(error));
    });
  });

  describe('Error logging', () => {
    beforeEach(() => {
      taskRunner = new TaskRunner({
        quietMode: false,
        parallelism: '1',
        changedProjectsOnly: false,
        terminal,
        allowWarningsInSuccessfulBuild: false
      });
    });

    const EXPECTED_FAIL: string = 'Promise returned by execute() resolved but was expected to fail';

    it('printedStderrAfterError', () => {
      taskRunner.addTask({
        name: 'stdout+stderr',
        isIncrementalBuildAllowed: false,
        execute: (writer: ITaskWriter) => {
          writer.write('Build step 1' + EOL);
          writer.writeError('Error: step 1 failed' + EOL);
          return Promise.resolve(TaskStatus.Failure);
        },
        hadEmptyScript: false
      });
      return taskRunner
        .execute()
        .then(() => fail(EXPECTED_FAIL))
        .catch(err => {
          expect(err.message).toMatchSnapshot();
          const allMessages: string = terminalProvider.getOutput();
          expect(allMessages).not.toContain('Build step 1');
          expect(allMessages).toContain('Error: step 1 failed');
          checkConsoleOutput(terminalProvider);
        });
    });

    it('printedStdoutAfterErrorWithEmptyStderr', () => {
      taskRunner.addTask({
        name: 'stdout only',
        isIncrementalBuildAllowed: false,
        execute: (writer: ITaskWriter) => {
          writer.write('Build step 1' + EOL);
          writer.write('Error: step 1 failed' + EOL);
          return Promise.resolve(TaskStatus.Failure);
        },
        hadEmptyScript: false
      });
      return taskRunner
        .execute()
        .then(() => fail(EXPECTED_FAIL))
        .catch(err => {
          expect(err.message).toMatchSnapshot();
          expect(terminalProvider.getOutput()).toMatch(/Build step 1.*Error: step 1 failed/);
          checkConsoleOutput(terminalProvider);
        });
    });

    it('printedAbridgedStdoutAfterErrorWithEmptyStderr', () => {
      taskRunner.addTask({
        name: 'large stdout only',
        isIncrementalBuildAllowed: false,
        execute: (writer: ITaskWriter) => {
          writer.write(`Building units...${EOL}`);
          for (let i: number = 1; i <= 50; i++) {
            writer.write(` - unit #${i};${EOL}`);
          }
          return Promise.resolve(TaskStatus.Failure);
        },
        hadEmptyScript: false
      });
      return taskRunner
        .execute()
        .then(() => fail(EXPECTED_FAIL))
        .catch(err => {
          expect(err.message).toMatchSnapshot();
          expect(terminalProvider.getOutput())
            .toMatch(/Building units.* - unit #1;.* - unit #3;.*lines omitted.* - unit #48;.* - unit #50;/);
          checkConsoleOutput(terminalProvider);
        });
    });

    it('preservedLeadingBlanksButTrimmedTrailingBlanks', () => {
      taskRunner.addTask({
        name: 'large stderr with leading and trailing blanks',
        isIncrementalBuildAllowed: false,
        execute: (writer: ITaskWriter) => {
          writer.writeError(`List of errors:  ${EOL}`);
          for (let i: number = 1; i <= 50; i++) {
            writer.writeError(` - error #${i};  ${EOL}`);
          }
          return Promise.resolve(TaskStatus.Failure);
        },
        hadEmptyScript: false
      });
      return taskRunner
        .execute()
        .then(() => fail(EXPECTED_FAIL))
        .catch(err => {
          expect(err.message).toMatchSnapshot();
          expect(terminalProvider.getOutput())
            .toMatch(/List of errors:\S.* - error #1;\S.*lines omitted.* - error #48;\S.* - error #50;\S/);
          checkConsoleOutput(terminalProvider);
        });
    });
  });

  describe('Warning logging', () => {
    describe('Fail on warning', () => {
      beforeEach(() => {
        taskRunner = new TaskRunner({
          quietMode: false,
          parallelism: '1',
          changedProjectsOnly: false,
          terminal,
          allowWarningsInSuccessfulBuild: false
        });
      });

      it('Logs warnings correctly', () => {
        taskRunner.addTask({
          name: 'success with warnings (failure)',
          isIncrementalBuildAllowed: false,
          execute: (writer: ITaskWriter) => {
            writer.write('Build step 1' + EOL);
            writer.write('Warning: step 1 succeeded with warnings' + EOL);
            return Promise.resolve(TaskStatus.SuccessWithWarning);
          },
          hadEmptyScript: false
        });

        return taskRunner
          .execute()
          .then(() => fail('Promise returned by execute() resolved but was expected to fail'))
          .catch(err => {
            expect(err.message).toMatchSnapshot();
            const allMessages: string = terminalProvider.getOutput();
            expect(allMessages).toContain('Build step 1');
            expect(allMessages).toContain('step 1 succeeded with warnings');
            checkConsoleOutput(terminalProvider);
          });
      });
    });

    describe('Success on warning', () => {
      beforeEach(() => {
        taskRunner = new TaskRunner({
          quietMode: false,
          parallelism: '1',
          changedProjectsOnly: false,
          terminal,
          allowWarningsInSuccessfulBuild: true
        });
      });

      it('Logs warnings correctly', () => {
        taskRunner.addTask({
          name: 'success with warnings (success)',
          isIncrementalBuildAllowed: false,
          execute: (writer: ITaskWriter) => {
            writer.write('Build step 1' + EOL);
            writer.write('Warning: step 1 succeeded with warnings' + EOL);
            return Promise.resolve(TaskStatus.SuccessWithWarning);
          },
          hadEmptyScript: false
        });

        return taskRunner
          .execute()
          .then(() => {
            const allMessages: string = terminalProvider.getOutput();
            expect(allMessages).toContain('Build step 1');
            expect(allMessages).toContain('Warning: step 1 succeeded with warnings');
            checkConsoleOutput(terminalProvider);
          })
          .catch(err => fail('Promise returned by execute() rejected but was expected to resolve'));
      });
    });
  });
});
