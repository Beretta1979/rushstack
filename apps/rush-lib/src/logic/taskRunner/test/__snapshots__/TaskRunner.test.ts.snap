// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskRunner Constructor throwsErrorOnInvalidParallelism 1`] = `"Invalid parallelism value of 'tequila', expected a number or 'max'"`;

exports[`TaskRunner Dependencies detectsDependencyCycle 1`] = `
"A cyclic dependency was encountered:
  foo
  -> bar
  -> foo
Consider using the cyclicDependencyProjects option for rush.json."
`;

exports[`TaskRunner Dependencies respectsDependencyOrder 1`] = `"Registered two[n]Registered one[n]Executing a maximum of 1 simultaneous processes...[-n-][n][x][37m[one] started[x][39m[n][x][32m1 of 2: [one] completed successfully in 0.00 seconds[x][39m[n][x][37m[two] started[x][39m[n][x][32m2 of 2: [two] completed successfully in 0.00 seconds[x][39m[n][n][x][32mSUCCESS (2)[x][39m[n][x][32m================================[x][39m[n][x][32mtwo (0.00 seconds)[x][39m[n][x][32mone (0.00 seconds)[x][39m[n][x][32m================================[-n-][x][39m[n][n]"`;

exports[`TaskRunner Dependencies respectsDependencyOrder 2`] = `""`;

exports[`TaskRunner Dependencies respectsDependencyOrder 3`] = `""`;

exports[`TaskRunner Dependencies respectsDependencyOrder 4`] = `""`;

exports[`TaskRunner Dependencies throwsErrorOnNonExistentDependency 1`] = `"The project 'bar' has not been registered."`;

exports[`TaskRunner Dependencies throwsErrorOnNonExistentTask 1`] = `"The task 'foo' has not been registered"`;

exports[`TaskRunner Error logging preservedLeadingBlanksButTrimmedTrailingBlanks 1`] = `"Project(s) failed to build"`;

exports[`TaskRunner Error logging preservedLeadingBlanksButTrimmedTrailingBlanks 2`] = `"Registered large stderr with leading and trailing blanks[n]Executing a maximum of 1 simultaneous processes...[-n-][n][x][37m[large stderr with leading and trailing blanks] started[x][39m[n][n][x][31mFAILURE (1)[x][39m[n][x][31m================================[x][39m[n][x][31mlarge stderr with leading and trailing blanks (0.00 seconds)[x][39m[n]List of errors:[-n-] - error #1;[-n-] - error #2;[-n-] - error #3;[-n-] - error #4;[-n-] - error #5;[-n-] - error #6;[-n-] - error #7;[-n-] - error #8;[-n-] - error #9;[-n-][...21 lines omitted...][-n-] - error #31;[-n-] - error #32;[-n-] - error #33;[-n-] - error #34;[-n-] - error #35;[-n-] - error #36;[-n-] - error #37;[-n-] - error #38;[-n-] - error #39;[-n-] - error #40;[-n-] - error #41;[-n-] - error #42;[-n-] - error #43;[-n-] - error #44;[-n-] - error #45;[-n-] - error #46;[-n-] - error #47;[-n-] - error #48;[-n-] - error #49;[-n-] - error #50;[n][x][31m================================[-n-][x][39m[n][n]"`;

exports[`TaskRunner Error logging preservedLeadingBlanksButTrimmedTrailingBlanks 3`] = `""`;

exports[`TaskRunner Error logging preservedLeadingBlanksButTrimmedTrailingBlanks 4`] = `""`;

exports[`TaskRunner Error logging preservedLeadingBlanksButTrimmedTrailingBlanks 5`] = `"[x][31m[-n-]1 of 1: [large stderr with leading and trailing blanks] failed to build![x][39m[n]"`;

exports[`TaskRunner Error logging printedAbridgedStdoutAfterErrorWithEmptyStderr 1`] = `"Project(s) failed to build"`;

exports[`TaskRunner Error logging printedAbridgedStdoutAfterErrorWithEmptyStderr 2`] = `"Registered large stdout only[n]Executing a maximum of 1 simultaneous processes...[-n-][n][x][37m[large stdout only] started[x][39m[n][n][x][31mFAILURE (1)[x][39m[n][x][31m================================[x][39m[n][x][31mlarge stdout only (0.00 seconds)[x][39m[n]Building units...[-n-] - unit #1;[-n-] - unit #2;[-n-] - unit #3;[-n-] - unit #4;[-n-] - unit #5;[-n-] - unit #6;[-n-] - unit #7;[-n-] - unit #8;[-n-] - unit #9;[-n-][...21 lines omitted...][-n-] - unit #31;[-n-] - unit #32;[-n-] - unit #33;[-n-] - unit #34;[-n-] - unit #35;[-n-] - unit #36;[-n-] - unit #37;[-n-] - unit #38;[-n-] - unit #39;[-n-] - unit #40;[-n-] - unit #41;[-n-] - unit #42;[-n-] - unit #43;[-n-] - unit #44;[-n-] - unit #45;[-n-] - unit #46;[-n-] - unit #47;[-n-] - unit #48;[-n-] - unit #49;[-n-] - unit #50;[n][x][31m================================[-n-][x][39m[n][n]"`;

exports[`TaskRunner Error logging printedAbridgedStdoutAfterErrorWithEmptyStderr 3`] = `""`;

exports[`TaskRunner Error logging printedAbridgedStdoutAfterErrorWithEmptyStderr 4`] = `""`;

exports[`TaskRunner Error logging printedAbridgedStdoutAfterErrorWithEmptyStderr 5`] = `"[x][31m[-n-]1 of 1: [large stdout only] failed to build![x][39m[n]"`;

exports[`TaskRunner Error logging printedStderrAfterError 1`] = `"Project(s) failed to build"`;

exports[`TaskRunner Error logging printedStderrAfterError 2`] = `"Registered stdout+stderr[n]Executing a maximum of 1 simultaneous processes...[-n-][n][x][37m[stdout+stderr] started[x][39m[n][n][x][31mFAILURE (1)[x][39m[n][x][31m================================[x][39m[n][x][31mstdout+stderr (0.00 seconds)[x][39m[n]Error: step 1 failed[n][x][31m================================[-n-][x][39m[n][n]"`;

exports[`TaskRunner Error logging printedStderrAfterError 3`] = `""`;

exports[`TaskRunner Error logging printedStderrAfterError 4`] = `""`;

exports[`TaskRunner Error logging printedStderrAfterError 5`] = `"[x][31m[-n-]1 of 1: [stdout+stderr] failed to build![x][39m[n]"`;

exports[`TaskRunner Error logging printedStdoutAfterErrorWithEmptyStderr 1`] = `"Project(s) failed to build"`;

exports[`TaskRunner Error logging printedStdoutAfterErrorWithEmptyStderr 2`] = `"Registered stdout only[n]Executing a maximum of 1 simultaneous processes...[-n-][n][x][37m[stdout only] started[x][39m[n][n][x][31mFAILURE (1)[x][39m[n][x][31m================================[x][39m[n][x][31mstdout only (0.00 seconds)[x][39m[n]Build step 1[-n-]Error: step 1 failed[n][x][31m================================[-n-][x][39m[n][n]"`;

exports[`TaskRunner Error logging printedStdoutAfterErrorWithEmptyStderr 3`] = `""`;

exports[`TaskRunner Error logging printedStdoutAfterErrorWithEmptyStderr 4`] = `""`;

exports[`TaskRunner Error logging printedStdoutAfterErrorWithEmptyStderr 5`] = `"[x][31m[-n-]1 of 1: [stdout only] failed to build![x][39m[n]"`;

exports[`TaskRunner Warning logging Fail on warning Logs warnings correctly 1`] = `"An error occurred."`;

exports[`TaskRunner Warning logging Fail on warning Logs warnings correctly 2`] = `"Registered success with warnings (failure)[n]Executing a maximum of 1 simultaneous processes...[-n-][n][x][37m[success with warnings (failure)] started[x][39m[n][n][x][33m[x][4mSUCCESS WITH WARNINGS (1)[x][24m[x][39m[n][x][33m================================[x][39m[n][x][33m[x][4msuccess with warnings (failure) (0.00 seconds)[x][24m[x][39m[n]Build step 1[-n-]Warning: step 1 succeeded with warnings[n][x][33m================================[-n-][x][39m[n][n]"`;

exports[`TaskRunner Warning logging Fail on warning Logs warnings correctly 3`] = `""`;

exports[`TaskRunner Warning logging Fail on warning Logs warnings correctly 4`] = `"[x][33m1 of 1: [success with warnings (failure)] completed with warnings in 0.00 seconds[x][39m[n][x][33mProject(s) succeeded with warnings[x][39m[n]"`;

exports[`TaskRunner Warning logging Fail on warning Logs warnings correctly 5`] = `""`;

exports[`TaskRunner Warning logging Success on warning Logs warnings correctly 1`] = `"Registered success with warnings (success)[n]Executing a maximum of 1 simultaneous processes...[-n-][n][x][37m[success with warnings (success)] started[x][39m[n][n][x][33m[x][4mSUCCESS WITH WARNINGS (1)[x][24m[x][39m[n][x][33m================================[x][39m[n][x][33m[x][4msuccess with warnings (success) (0.00 seconds)[x][24m[x][39m[n]Build step 1[-n-]Warning: step 1 succeeded with warnings[n][x][33m================================[-n-][x][39m[n][n]"`;

exports[`TaskRunner Warning logging Success on warning Logs warnings correctly 2`] = `""`;

exports[`TaskRunner Warning logging Success on warning Logs warnings correctly 3`] = `"[x][33m1 of 1: [success with warnings (success)] completed with warnings in 0.00 seconds[x][39m[n]"`;

exports[`TaskRunner Warning logging Success on warning Logs warnings correctly 4`] = `""`;
